Add contributor form into app

aapeliv opened the pull request:
There's a shared component, which is embedded on the dashboard. The backend tracks whether logged in users have filled in the form or not. Form submissions are sent to an AWS lambda function. I've also added a non-logged in route `/contribute` for those who have not signed up (for whatever reason?). Several form inputs are hidden and filled in automatically if the user is signed in.

* ~~Itsi would like the radio group to be turned into a set of checkboxes, but that's a lot of effort~~
* ~~Add a separate username field + lambda func tweaks~~
* ~~Would be good to get the user's email into the form so the form backend can email the person~~
* Needs some tweaks in the AWS lambda functions: email needs to be updated to reflect current situation, and emails to us need to be updated to include username.

It'd be great to eventually move the lambda function concoction into the backend, but that can come later.

aapeliv committed:
Add contributor form into app

aapeliv committed:
Add contributor form RPC call and model stuff

aapeliv committed:
Update contributor form to grab data from dedicated function

aapeliv commented:
@kalvdans: could you please review backend bit, it's tiny :)

aapeliv committed:
Clean up contributor form

aapeliv committed:
Turn contributor radios into checkboxes

aapeliv commented:
I've done an awful hack to get checkboxes and have them turned nicely into a comma-delimited string...

kalvdans reviewed (dismissed):
backend looks solid

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+  useEffect(() => {
+    if (authState.authenticated) {
+      (async () => {
+        const info = await service.account.getContributorFormInfo();
```
It's probably easier and safer to do this with React Query, since it's easy for this kinda async stuff to run into the setting state on unmounted component warning (happens whenever some async stuff is still loading but user navigates away from page before it finishes loading) that could cause memory leaks. React Query don't have this issue since it's global outside of React.

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+        setValue("email", info.email);
+        setValue("age", info.age);
+        setValue("gender", info.gender);
+        setValue("location", info.location);
```
Don't these fields also exist in the form? It's better to set them as `defaultValue` on the form fields instead, otherwise React hook form warns about this

aapeliv replied:
I tried for quite a while to get this to work, but then couldn't figure out how to get these into the form without `setValue`, though I've moved stuff around

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+        setError("An unknown error occured");
+      }
+    } catch (e) {
+      setError(e.message);
```
Use mutation from React Query, again handles all of this error and loading state so slightly less boilerplate. You can more or less take the bit in the try-block and put it in the mutation but without needing the try-catch and setState/error stuff (react query catches error for us when things go wrong).

aapeliv replied:
Thanks, it is quite a lot easier!

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+                    rowsMax={6}
+                  />
+                  <Button
+                    classes={{}}
```
```suggestion
```

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+  };
+
+  return (
+    <>
```
This outer fragment `<>` is unnecessary since it's a single ternary

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+        <>
+          {filled ? (
+            <>
+              <p>{ALREADY_FILLED_IN}</p>
```
Use `Typography` so it uses styles from the theme

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+            <>
+              {success ? (
+                <>
+                  <p>{SUCCESS_MSG}</p>
```
Use `Typography`

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+                  <p>{SUCCESS_MSG}</p>
+                  {!authState.authenticated && (
+                    <>
+                      <p>{PLEASE_SIGN_UP}</p>
```
Same as above

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+                </>
+              ) : (
+                <form onSubmit={submit}>
+                  <p>{FILL_IN_THE_FORM}</p>
```
Use `Typography`

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+                      <Button component={Link} to={signupRoute}>
+                        {SIGN_UP}
+                      </Button>
+                      <p>{YOU_CAN_ALSO}</p>
```
Use `Typography`

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+                        className={classNames({
+                          [classes.hidden]: authState.authenticated,
+                        })}
+                      />
```
Better to conditionally not render this if user is authenticated rather than hiding it

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+                    render={({ onChange, value }) => (
+                      <FormControl>
+                        <FormLabel>{CONTRIBUTE_LABEL}</FormLabel>
+                        <FormGroup aria-label="contribute">
```
Put the aria-label in constant file too

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+                      </FormControl>
+                    )}
+                  />
+                  <p>{QUESTIONS_OPTIONAL}</p>
```
Use `Typography`

darrenvong commented on app/frontend/src/features/Contribute.tsx:
```diff
@@ -0,0 +1,98 @@
+    height: "100%",
+    width: "100%",
+    backgroundColor: theme.palette.common.white,
+    // backgroundAttachment: "scroll, local",
```
```suggestion
    // backgroundAttachment: "scroll, local",
```
Remove if we don't need this?

darrenvong commented on app/frontend/src/features/Contribute.tsx:
```diff
@@ -0,0 +1,98 @@
+    display: "flex",
+  },
+  link: {
+    borderRadius: theme.shape.borderRadius / 3,
```
```suggestion
    borderRadius: theme.shape.borderRadius,
```
I've changed `borderRadius` value in the theme recently if you rebase with `develop` so don't need to divide by 3 anymore

darrenvong commented on app/frontend/src/features/Contribute.tsx:
```diff
@@ -0,0 +1,98 @@
+  },
+  formWrapper: {
+    backgroundColor: theme.palette.background.default,
+    borderRadius: theme.shape.borderRadius / 3,
```
```suggestion
    borderRadius: theme.shape.borderRadius,
```

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+    contribute: string,
+    setContribute: (s: string) => void
+  ) => {
+    const currentChoices = contribute.split(",").filter((v) => !!v);
```
What's wrong with the original form value that comes from react hook form's `onChange`? Couldn't the individual checkbox values be collected in `onSubmit` instead just before it gets sent to the fetch call rather than on every change?

aapeliv replied:
I have to do the filtering/concat on every change anyway, so I thought it'd be easier to just split/join here as well

darrenvong replied:
Hmm yeah even doing it on submit is gonna be hacky and this is likely a one time thing people use, so happy to leave this as it is as this is still more readable than what I initially had in mind

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+                          </FormControl>
+                        )}
+                      />
+                      <TextField
+                        inputRef={register}
+                        id="location"
+                        margin="normal"
+                        name="location"
+                        label={LOCATION_LABEL}
+                        helperText={LOCATION_HELPER}
+                        fullWidth
+                      />
```
```suggestion
                      <TextField
                        inputRef={register}
                        id="location"
                        margin="normal"
                        name="location"
                        label={LOCATION_LABEL}
                        helperText={LOCATION_HELPER}
                        fullWidth
                      />
```
Set the `defaultValue` prop on this (got a react-hook-form warning on this) with the result of `getContributorFormInfo`, should be accessible if using react query to do this once loaded.

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,384 @@
+              <Button onClick={() => setFilled(false)}>{FILL_IN_AGAIN}</Button>
+            </>
+          ) : (
+            <>
```
Don't need this fragment either - single ternary again

darrenvong reviewed (changes requested):
Hopefully I caught most things

aapeliv committed:
Fix easy review comments

aapeliv committed:
Use react query for initial contributor form info query

aapeliv committed:
Refactor contributor form to user react query

aapeliv commented:
@darrenvong: I rebased on `develop`, fixed up the easy comments, and then refactored to use react query.

I was still having trouble setting the results from the first query (contributor form info endpoint) into the form, so I've still got it as hidden when logged in and using `setValue` in `onSuccess` in the react query query

darrenvong commented on app/frontend/src/features/contributorForm/ContributorForm.tsx:
```diff
@@ -0,0 +1,367 @@
+        body: JSON.stringify(data),
+      }
+    );
+    return (await response.json()) as FormResponse;
```
```suggestion
    return response.json();
```
Unnecessary `await` if you are already returning a Promise - the caller (React Query) will await it correctly.

darrenvong reviewed (changes requested):
Very minor thing left, but this is a really great refactor overall 🎉

aapeliv committed:
Clean up

darrenvong reviewed (approved):
Assume build succeeds as before
